### PR TITLE
Vibhatsu/signals n slots

### PR DESCRIPTION
**Pull request by Demigod345**

### CODE REVIEW

File: src/audio.cpp [hunk 1]
```diff
@@ -230,11 +224,29 @@ void APU::clearRegisters()
...
 
+// Write on Memory Write
+void APU::onWrite(Word address, Byte value)
+{
+	writeUpdate(address, value, true);
+	writeUpdate(0xFF26, 0xFF);
+}
+
+// Write Update
+void APU::writeUpdate(Word address, Byte value, bool MemWrite)
```
Comment (Demigod345) [inline src/audio.cpp]: remove redundant code

File: src/audio.h [hunk 2]
```diff
@@ -172,4 +172,10 @@ class APU
 	void stepAPU(int cycles);
 	void clearRegisters();
 	void setMemoryMap(MemoryMap* map) { mMap = map; }
+	// Writes back on  Memory Write
+	void onWrite(Word address, Byte value);
+	// Write update
+	void writeUpdate(Word address, Byte value, bool WriteMem = false);
```
Comment (Demigod345) [inline src/audio.h]:
```
enum {
APU_FLAG_WRITE_BACK = 0,
APU_FLAG_WRITE_TO_MEMORY
};

writeUpdate (address, Flag);
```

File: src/audio.cpp [hunk 3]
```diff
@@ -55,6 +55,13 @@ void APU::test()
 	printf("APU test\n");
 }
 
+// Connecting CallBack function
+void APU::setSignalCallback()
+{
+	if (mMap)
+		mMap->connectObserver([this](Word address, Byte value) { this->onWrite(address, value); });
```
Comment (Demigod345) [inline src/audio.cpp]: use descriptive naming

File: src/audio.cpp [hunk 4]
```diff
@@ -55,6 +55,13 @@ void APU::test()
 	printf("APU test\n");
 }
 
+// initialize the writehandler of memorymap
+void APU::initializeWriteHandler()
+{
+	if (mMap)
+		mMap->setAduioWriteHandler([this](Word address) { this->onWrite(address); });
```
Comment (Demigod345) [inline src/audio.cpp]: spelling mistake here

Comment (Demigod345) [reply src/audio.cpp]: give more descriptive name to `onWrite ()`

File: src/audio.cpp [hunk 5]
```diff
@@ -207,8 +202,7 @@ void APU::stepAPU(int cycles)
 		Word address[] = { 0xFF19, 0xFF1E, 0xFF23, 0xFF26 };
 		for (auto addr : address)
 		{
-			Byte reg = readByte(addr);
-			mMap->writeBackMemory(addr, reg);
+			writeUpdate(addr, AUDIO_WRITE);
```
Comment (Demigod345) [inline src/audio.cpp]: same better naming

File: src/audio.h [hunk 6]
```diff
@@ -2,7 +2,9 @@
 #include "types.h"
 #include "mmap.h"
 #include <stdio.h>
-#include <SDL.h>
+#include <SDL.h> // SDL Audio
+
+#define AUDIO_CHANNEL_CONTROL 0xFF26
```
Comment (Demigod345) [inline src/audio.h]: `AUDIO_MASTER_CONTROL_REGISTER` would be better.

File: src/audio.h [hunk 7]
```diff
@@ -12,6 +14,14 @@ enum Channel
 	CH4 = 3
 };
 
+// For checking whether Write is due to memory or APU itself
+enum audioWriteFlag
+{
+	AUDIO_WRITE = 0,
+	AUDIO_MEMORY_WRITE = 1
+
```
Comment (Demigod345) [inline src/audio.h:22]: space

Comment (Demigod345) [reply src/audio.h:22]: generally enums/user-defined classes/structs are Pascal Case

File: src/mmap.cpp [hunk 8]
```diff
@@ -1,6 +1,8 @@
 #include "mmap.h"
 #include <cstring>
 
+// std::function<void(Word, Byte)> globalFunction = nullptr;
```
Comment (Demigod345) [inline src/mmap.cpp]: remove commented code

File: src/mmap.h [hunk 9]
```diff
@@ -1,7 +1,8 @@
 #pragma once
 #include "types.h"
 #include <stdio.h>
-#include <queue>
+#include <functional>
+// extern std::function<void(Word, Byte)> globalFunction;
```
Comment (Demigod345) [inline src/mmap.h]: remove comments

**Review by Demigod345 — COMMENTED** [review]
Nice changes overall, just left some coding style nits.

File: src/audio.cpp (hunk 4 above)
Comment (v1bh475u) [reply src/audio.cpp]: How about `onMemoryWrite()`

File: src/audio.cpp (hunk 5 above)
Comment (v1bh475u) [reply src/audio.cpp]: how about `audioRegisterUpdate`?

**Pull request merged**